File: 09/02/pages/BasePage.js
'use strict';

var driver,
    Promise = require('selenium-webdriver').promise;

function BasePage(driver) {
  this.driver = driver;
}

BasePage.prototype.visit = function(url) {
  this.driver.get(url);
};

BasePage.prototype.find = function(locator) {
  return this.driver.findElement(locator);
};

BasePage.prototype.click = function(locator) {
  this.find(locator).click();
};

BasePage.prototype.type = function(locator, inputText) {
  this.find(locator).sendKeys(inputText);
};

BasePage.prototype.isDisplayed = function(locator) {
  var defer = Promise.defer();
  this.find(locator).isDisplayed().then(function(isDisplayed) {
    defer.fulfill(isDisplayed);
  }, function(error) {
    if (error.name === 'NoSuchElementError') {
      defer.fulfill(false);
    } else {
      defer.reject(error);
    }
  });
  return defer.promise;
};

module.exports = BasePage;

File: code-examples/ruby/16/01/pages/base_page.rb
require_relative '../spec/config'

class BasePage
  include Config

  def initialize(driver)
    @driver = driver
  end

  def visit(url)
    if url.start_with? 'http'
      @driver.get url
    else
      @driver.get config[:base_url] + url
    end
  end

  def find(locator)
    @driver.find_element locator
  end

  def click(locator)
    find(locator).click
  end

  def type(text, locator)
    find(locator).send_keys text
  end

  def is_displayed?(locator)
    begin
      return find(locator).displayed?
    rescue Selenium::WebDriver::Error::NoSuchElementError
      return false
    end
  end

  def wait_for(seconds = 15)
    Selenium::WebDriver::Wait.new(timeout: seconds).until { yield }
  end

end
